algebras/algebra/src/main/scala/endpoints4s/algebra/ChunkedEntities.scala

```diff
@@ -77,24 +93,56 @@ trait ChunkedEntities extends EndpointsWithCustomErrors {
   *
   * @group algebras
   */
-trait ChunkedJsonEntities extends ChunkedEntities with JsonCodecs {
+trait ChunkedJsonEntities extends ChunkedJsonRequestEntities with ChunkedJsonResponseEntities
+
+trait ChunkedJsonRequestEntities
+    extends ChunkedRequestEntities
+    with JsonCodecs
+    with RequestChunkCodec {
 
   /** A request entity carrying chunks of JSON values
     *
     * @tparam A Type of values serialized into JSON
     * @group operations
     */
-  def jsonChunksRequest[A](implicit
+  def jsonChunksRequest[A](chunkCodec: RequestChunkCodec)(implicit
       codec: JsonCodec[A]
   ): RequestEntity[Chunks[A]]
+}
+
+trait ChunkedJsonResponseEntities
+    extends ChunkedResponseEntities
+    with JsonCodecs
+    with ResponseChunkCodec {
 
   /** A response entity carrying chunks of JSON values
     *
     * @tparam A Type of values serialized into JSON
     * @group operations
     */
-  def jsonChunksResponse[A](implicit
+  def jsonChunksResponse[A](chunkCodec: ResponseChunkCodec)(implicit
       codec: JsonCodec[A]
   ): ResponseEntity[Chunks[A]]
+}
+
+trait RequestChunkCodec {
+
+  type RequestChunkCodec
+}
+
+trait ResponseChunkCodec {
+
+  type ResponseChunkCodec
+}
+
+trait NewLineChunkCodec extends NewLineRequestChunkCodec with NewLineResponseChunkCodec
+
+trait NewLineRequestChunkCodec extends RequestChunkCodec {
+
+  def newLineRequestChunkCodec[A]: RequestChunkCodec
+}
+
+trait NewLineResponseChunkCodec extends ResponseChunkCodec {
```

Comment:
There are lots of traits introduced in this file. We should be careful not to introduce too many traits because they create complexity at the use-site. So, we always need to assess whether a trait is really needed or not. In general, a good reason to introduce a separate trait is to support interpreters that can’t implement one part of the algebra. Is this the case here? Do we have examples of interpreters that can implement `ChunkedRequestEntities` but can’t implement `ChunkedResponseEntities` (or _viceversa_)? Regarding `ChunkCodec`, the fact that a whole trait defines the delimiter might be complicated to use: if someone wants to use different delimiters in the same service, they will have to create two services (more or less).

Reply:
So, to reply to my own question: yes, browsers can’t implement `ChunkedRequestEntities` but they can implement `ChunkedResponseEntities`.

Now, regarding the delimiters, I wonder if it would be possible to introduce a `Delimiters` algebra, extended by both `ChunkedJsonRequestEntities` and `ChunkedJsonResponseEntities`:

~~~ scala
trait Delimiters {
  type Delimiter
  def newLineDelimiter: Delimiter
}
~~~

(something along those lines)

Reply:
Just a note in regards to naming - this abstraction is more than just about the delimiters, it can be very large arrays or length prefixes too, so that is why I chose the name _codec_. It saw it also called by _framing_. Are you sure about naming this a delimiter?

Is this abstraction not more limiting? Because you are forcing interpreters to implement `newLineDelimiter` if they want to deal with `Delimiter`. That is the reason why I split it off - to allow interpreters full flexibility.

Reply:
This `Delimiter` (_codec_, _framing_) is basically a flow that transforms the `Chunks`. If you could take a look at my provided implementations, the type definitions sort of tie together naturally.

Reply:
That being said, and just to clarify, I do not think that algebras should tie chunks and their transformations together.

Reply:
After looking at the implementation of the interpreters, it seems that we always implement the newline frame delimiter. I don’t think we should provide too many options in the algebra, unless there is a really good reason to do so.

It would be nice if it was possible to perform JSON streaming with very little code like:

~~~ scala
trait StreamingJson extends algebra.Endpoints with algebra.ChunkedJsonEntities {
  val foo = endpoint(get(path / "foo"), chunkedJsonResponse[Foo](framing = newLineDelimiterFraming))
}

object Documentation extends StreamingJson with openapi.Endpoints with openapi.ChunkedJsonEntities {
  val api = openApi(...)(foo) 
}

object Client extends StreamingJson with client.Endpoints with client.ChunkedJsonEntities

object Server extends StreamingJson with server.Endpoints with server.ChunkedJsonEntities
~~~